api/client/commands.go

```diff
@@ -162,21 +163,19 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 			}
 
 			// Verify that 'filename' is within the build context
-			if !strings.HasSuffix(absRoot, string(os.PathSeparator)) {
-				absRoot += string(os.PathSeparator)
-			}
-			if !strings.HasPrefix(filename, absRoot) {
-				return fmt.Errorf("The Dockerfile (%s) must be within the build context (%s)", *dockerfileName, root)
+			filename, err = symlink.FollowSymlinkInScope(filename, absRoot)
```

Comment:
So `Dockerfile` being a link to a file outside of the context is a case where `docker build .` and `docker build -f Dockerfile .` will behave differently.

It's tricky.